### ISSUE TITLE
automatically upload important artifacts to a GCP bucket

### ISSUE THREAD

**Issue by bhearsum** [mozilla/translations#466]
Everything we run in Taskcluster expires eventually. Anything we care about retaining must be uploaded elsewhere, and to ensure it happens we should automate it. We have existing tools for doing this pretty easily in Taskcluster.

@eu9ene or @gregtatum - can we get a list of artifacts from each pipeline step that we should be uploading? Models coming out of training tasks are the obvious thing - but I imagine there might be other things too.

**Comment by eu9ene** [comment]
Related to https://github.com/mozilla/firefox-translations-training/issues/191

We should upload the artifacts and logs for:
- all the `train-` steps
- `vocab`
- `export`
-  all the `evaluate-` steps
- train action config

I don't think I inspected the datasets or other artifacts for the intermediate steps of the old experiments.   

Ideally, it should follow the existing directory structure that we have for the uploaded snakemake experiments in the bucket but you might want to come up with a different structure that's easier to manage as an archive.

The benefit of having the same structure would be that we can reupload the experiments to W&B if we want to change something in the uploading process or add more metadata but logs and artifacts have already expired in a Taskcluster task.

**Comment by eu9ene** [comment]
Also, do we want to move the data to a more production-grade bucket from `gs://releng-translations-dev`?

**Comment by bhearsum** [comment]
> Also, do we want to move the data to a more production-grade bucket from `gs://releng-translations-dev`?

Yeah, I think that would be good. I'd like to upload the artifacts from https://github.com/mozilla/firefox-translations-training/issues/312#issuecomment-1946802157 to a production place, as well as any new uploads that happen automatically.

**Comment by eu9ene** [comment]
It's basically models, logs and experiments from the [snakemake directory structure](https://mozilla.github.io/firefox-translations-training/snakemake.html#directory-structure):
```
gsutil ls gs://releng-translations-dev
gs://releng-translations-dev/data/
gs://releng-translations-dev/experiments/
gs://releng-translations-dev/logs/
gs://releng-translations-dev/models/
```

We use `/data` to store custom datasets, unlike for snakemake, but we can always rename this one as needed.

Also, we thought [here](https://github.com/mozilla/firefox-translations-training/issues/312#issuecomment-1888034167) on how to name the experiments (task groups) from Taskcluster and so far came up with `<experiment_name_from_config>_<task_group>` because multiple task groups can be launched with the same experiment name. It's not ideal because, for example, it's hard to say when the experiment was launched and we might want to add a date there too or handle it differently by adding more metadata to W&B. Anyway it's better to be consistent and name the folders in the same way on GCP.

And here is the [list](https://github.com/mozilla/firefox-translations-training/issues/312#issuecomment-1946802157) of the groups we have so far.

I think when we get to W&B model registry the underlying storage will be on GCP anyway and W&B will store only additional metadata so we'll benefit from consistency of the experiments structure across W&B projects and GCP. We'll likely use [this API](https://docs.wandb.ai/guides/model_registry/log-model-to-experiment) to track the GCP model artifacts on W&B.

**Comment by bhearsum** [comment]
So to spell it out a bit more, if our experiment name was `retrain`, and we were doing ru-en, we would expect the following:
* `train-backwards` artifacts in `/models/ru-en/retrain/backward`
* `train-teacher` artifacts in `/models/ru-en/retrain/teacher0` (and maybe additional directories if ensemble > 1)
* `train-student` artifacts in `/models/ru-en/retrain/student`
* `finetune-student` artifacts in `/models/ru-en/retrain/student-finetuned`
* `train-vocab` artifacts in `/models/ru-en/retrain/vocab` ? (not sure about this...it's not called out in the snakemake doc)
* `evaluate-backwards` artifacts in `/models/ru-en/retrain/evaluation/backward`
* `evaluate-teacher-ensemble` artifacts in `/models/ru-en/retrain/evaluation/teacher-ensemble`
* (etc. for the other training steps)
* `export` artifacts in `/models/ru-en/retrain/exported`
* logs for all of the above in the same directories, except `/logs` instead of `/models` at the start
* train action config in `/experiments/ru-en/retrain/config.json` ? (IIRC, the old config was sh, and we now have json)

**Comment by bhearsum** [comment]
I'm still working on downloading everything, but in the meantime, does this partial list of destinations look sensible? (The second column is where they'd end up in GCS - the first column is just where I have them locally at the moment.)

**Comment by eu9ene** [comment]
Yes, it looks correct overall, some additions:
- let's not forget about `quantize` and `evaluate quantized`, they also produce the model and evaluation results
- instead of `retrain` let's use `retrain_<task_group_id>` because we'll have multiple runs with the same experiment name and want to be consistent with W&B naming

**Comment by eu9ene** [comment]
The old config is yaml, so let's upload the yaml from Taskcluster instead of json. Old en-ru experiment: `gs://releng-translations-dev/experiments/en-ru/ny-retraining/config.yml`

Both train.log and live.log are useful.

**Comment by eu9ene** [comment]
> > Vocab is also needed even though we now also store it as model artifacts.
> 
> And it belongs in directories like `models/en-ru/retrain1_<task group_id>/vocab` ?

Correct

> > ```
> > * instead of `retrain` let's use `retrain_<task_group_id>` because we'll have multiple runs with the same experiment name and want to be consistent with W&B naming
> > ```
> 
> How does this work when we have multiple task groups for the same experiment? Do we end up with multiple directories per experiment?

Yes, we don't enforce unique experiment name per train action, so those can be different experiments. This is the approach we're taking in W&B dashboards

> I did include evaluate-quantized - I assumed that was what ended up in the speed directory - is that wrong?

Yes, we call both the model and evaluation results `speed` in the legacy directory structure. One option is to keep it this way for consistency. Alternatively, we could call it `quantized` to match the step names and the names we use on W&B dashboard. We would likely need to change W&B uploader to support this. I would prefer the latter option.

**Comment by bhearsum** [comment]
> > I did include evaluate-quantized - I assumed that was what ended up in the speed directory - is that wrong?
> 
> Yes, we call both the model and evaluation results `speed` in the legacy directory structure. One option is to keep it this way for consistency. Alternatively, we could call it `quantized` to match the step names and the names we use on W&B dashboard. We would likely need to change W&B uploader to support this. I would prefer the latter option.

No opinion from me - just let me know what you want to do.

**Comment by eu9ene** [comment]
> > > I did include evaluate-quantized - I assumed that was what ended up in the speed directory - is that wrong?
> > 
> > 
> > Yes, we call both the model and evaluation results `speed` in the legacy directory structure. One option is to keep it this way for consistency. Alternatively, we could call it `quantized` to match the step names and the names we use on W&B dashboard. We would likely need to change W&B uploader to support this. I would prefer the latter option.
> 
> No opinion from me - just let me know what you want to do.

Ok, let's name them both `quantized` instead of `speed`.

**Comment by bhearsum** [comment]
For lack of a better place to track it - I just finished uploading all of the artifacts from [these task groups](https://github.com/mozilla/firefox-translations-training/issues/312#issuecomment-1946802157) with a set of hacky scripts.

I'm planning to prioritize the automatic uploads this issue is tracking ASAP. (I imagine we'll need another round of manual uploads once the automated ones are completed, though.)
